In [2]:
import numpy as np
from scipy.stats import norm

<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### Q1. Two-Proportion Z-Test

**Control:** 180 conversions / 2,500 users &nbsp;|&nbsp; **Treatment:** 210 conversions / 2,500 users

Perform a full two-proportion z-test. Compute the pooled proportion, standard error, z-statistic, and p-value. State your conclusion at **α = 0.05**.

</div>

We have two independent groups and a binary outcome (converted / not), so a two-proportion z-test is appropriate.

Under H₀ (p₁ = p₂), we pool both samples to estimate a single common proportion, then measure how far the observed difference is in standard error units.

| Step | Formula |
|------|---------|
| Pooled proportion | $\hat{p} = \dfrac{x_1 + x_2}{n_1 + n_2}$ |
| Standard error | $SE = \sqrt{\hat{p}(1-\hat{p})\left(\dfrac{1}{n_1}+\dfrac{1}{n_2}\right)}$ |
| Z-statistic | $z = \dfrac{\hat{p}_2 - \hat{p}_1}{SE}$ |
| p-value (one-tailed) | `norm.sf(z)` |

**Hypotheses:**  
H₀: p₁ = p₂ (no difference in conversion rates)  
H₁: p₂ > p₁ (treatment converts better)

In [8]:
# Given data
x1 = 180   # conversions in control
n1 = 2500  # users in control
x2 = 210   # conversions in treatment
n2 = 2500  # users in treatment
alpha = 0.05

In [4]:
# Individual conversion rates
p1 = x1/n1
p2 = x2/n2

In [5]:
# Pooled proportion: best estimate of p under H0
p_pool = (x1+x2)/(n1+n2)
p_pool

0.078

In [6]:
# Standard error of the difference in proportions
se = np.sqrt(p_pool * (1-p_pool) * (1/n1 + 1/n2))
se

np.float64(0.007585037903662711)

In [7]:
# Z-statistic: how many SEs is the observed difference from 0
z_test = (p2-p1)/se
z_test

np.float64(1.5820619688934416)

In [9]:
# Critical value for one-tailed test at alpha = 0.05
z_crit = norm.ppf(1-alpha)
z_crit

np.float64(1.6448536269514722)

In [10]:
# p-value: P(Z > z_test) under H0  |  sf = survival function = 1 - CDF
pval = norm.sf(z_test)
pval

np.float64(0.05681771225610912)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Conclusion**

z_test (1.582) < z_crit (1.645) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.057) > alpha (0.05)

**Fail to reject H₀.** The 30-conversion lift is not statistically significant at 5% level. The treatment shows a trend but the evidence is insufficient to conclude a real difference.

</div>